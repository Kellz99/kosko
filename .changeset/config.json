{
  "$schema": "https://unpkg.com/@changesets/config@1.5.0/schema.json",
  "changelog": ["@changesets/changelog-github", { "repo": "tommy351/kosko" }],
  "commit": false,
  "linked": [["@kosko/cli", "kosko"]],
  "access": "public",
  "baseBranch": "master",
  "updateInternalDependencies": "patch",
  "ignore": []
}
