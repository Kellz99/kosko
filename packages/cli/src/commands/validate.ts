import { Command, getLogger } from "../cli/command";
import {
  BaseGenerateArguments,
  generateBuilder,
  generateHandler
} from "./generate";

export type ValidateArguments = BaseGenerateArguments;

export const validateCmd: Command<ValidateArguments> = {
  command: "validate [components..]",
  describe: "Validate components",
  builder(argv) {
    /* istanbul ignore next */
    return generateBuilder(argv)
      .example("$0 validate", "Validate components")
      .example("$0 validate foo bar", "Specify components")
      .example("$0 validate foo_*", "Use glob syntax");
  },
  async handler(args) {
    const logger = getLogger(args);
    await generateHandler({ ...args, validate: true });
    logger.success("Components are valid");
  }
};
