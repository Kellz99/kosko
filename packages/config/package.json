{
  "name": "@kosko/config",
  "version": "3.0.0",
  "description": "Load kosko configuration.",
  "homepage": "https://kosko.dev",
  "repository": "https://github.com/tommy351/kosko/tree/master/packages/config",
  "author": "Tommy Chen <tommy351@gmail.com>",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "files": [
    "/dist/"
  ],
  "keywords": [
    "kosko"
  ],
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=14.18.0"
  },
  "dependencies": {
    "@iarna/toml": "^2.2.5",
    "@kosko/common-utils": "workspace:^",
    "@kosko/log": "workspace:^",
    "superstruct": "^0.15.3",
    "tslib": "^2.4.0"
  },
  "devDependencies": {
    "@kosko/jest-preset": "workspace:^"
  },
  "jest": {
    "preset": "@kosko/jest-preset"
  }
}
