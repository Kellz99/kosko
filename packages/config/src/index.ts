/**
 * @packageDocumentation
 * @module @kosko/config
 */

export * from "./config";
export { EnvironmentConfig, Config } from "./types";
export * from "./validate";
