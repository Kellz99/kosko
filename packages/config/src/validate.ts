import { ErrorObject } from "ajv";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const ajvValidate = require("../dist/ajv-validate");

function formatErrors(errors: ErrorObject[]): string {
  if (!errors.length) return "No errors";
  return errors.map((err) => `${err.dataPath} ${err.message}`).join(", ");
}

export class ValidationError extends Error {
  public constructor(public readonly errors: ErrorObject[]) {
    super(formatErrors(errors));
  }
}

ValidationError.prototype.name = "ValidationError";

/**
 * Validates data with kosko configuration schema. It throws a ValidationError
 * when validation failed.
 *
 * @param data
 */
export function validate(data: unknown): void {
  const valid = ajvValidate(data);
  if (valid) return;

  throw new ValidationError(ajvValidate.errors || []);
}
