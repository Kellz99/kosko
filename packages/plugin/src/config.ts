import { Struct, StructError, validate } from "superstruct";

/**
 * @public
 */
export class ConfigError extends Error {
  public constructor(err: StructError) {
    super("Invalid config");

    let stack = `${this.name}: ${this.message}`;

    for (const failure of err.failures()) {
      stack += `\n- ${failure.message}`;
    }

    this.stack = stack;
  }
}

ConfigError.prototype.name = "ConfigError";

/**
 * Validate a value against a schema.
 *
 * @public
 *
 * @throws ConfigError
 * Thrown if the value is invalid.
 */
export function validateConfig<T>(value: unknown, schema: Struct<T>): T {
  const [err, config] = validate(value, schema);
  if (!err) return config;

  throw new ConfigError(err);
}
