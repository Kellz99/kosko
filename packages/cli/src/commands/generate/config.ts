import { getConfig, searchConfig } from "@kosko/config";
import { CLIError } from "../../cli/error";
import { BaseGenerateArguments } from "./types";

export async function loadConfig(args: BaseGenerateArguments) {
  const base = await searchConfig(args.cwd);
  const {
    components = [],
    require = [],
    loaders = []
  } = args.env ? getConfig(base, args.env) : base;
  const config = {
    ...base,
    components:
      args.components && args.components.length ? args.components : components,
    require: [...require, ...(args.require || [])],
    loaders: [...loaders, ...(args.loader || [])]
  };

  if (!config.components.length) {
    throw new CLIError("No components are given", {
      output:
        "No components are given. Set components in a config file or in arguments."
    });
  }

  return config;
}
