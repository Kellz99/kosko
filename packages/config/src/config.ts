import toml from "@iarna/toml";
import Debug from "debug";
import fs from "fs";
import { join } from "path";
import { promisify } from "util";
import { Config, EnvironmentConfig } from "./types";
import { validate } from "./validate";

const debug = Debug("kosko:config");
const readFile = promisify(fs.readFile);

/**
 * Parses and validates a config file from the specified path.
 *
 * @param path Path of the config file.
 */
export async function loadConfig(path: string): Promise<Config> {
  const content = await readFile(path, "utf8");
  const data = await toml.parse.async(content);

  debug("Found config at", path);
  return validate(data);
}

/**
 * Searchs config files in the specified directory. Returns an empty object when
 * config files does not exist in the directory.
 *
 * @param cwd Path to the working directory.
 */
export async function searchConfig(
  cwd: string = process.cwd()
): Promise<Config> {
  const path = join(cwd, "kosko.toml");

  try {
    return await loadConfig(path);
  } catch (err) {
    if (err.code === "ENOENT") return {};

    debug("Config load failed", err);
    throw err;
  }
}

function flatten<T>(...arrays: (ReadonlyArray<T> | undefined)[]): T[] {
  return arrays.reduce((acc = [], item = []) => acc.concat(item), []) as T[];
}

/**
 * Returns environment configs merged with global configs.
 *
 * @param config Config object.
 * @param env Environment name.
 */
export function getConfig(
  config: Config,
  env: string
): Required<EnvironmentConfig> {
  const { environments = {} } = config;
  const envConfig = environments[env] || {};

  return {
    require: flatten(config.require, envConfig.require),
    components: flatten(config.components, envConfig.components)
  };
}
