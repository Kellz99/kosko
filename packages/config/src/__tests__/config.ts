import { join } from "node:path";
import { getConfig, loadConfig, searchConfig } from "../config";
import { ValidationError } from "../validate";

const fixturePath = join(__dirname, "..", "__fixtures__");

describe("loadConfig", () => {
  describe("when file exists", () => {
    test("should load the config", async () => {
      const config = await loadConfig(join(fixturePath, "toml", "kosko.toml"));
      expect(config).toMatchSnapshot();
    });
  });

  describe("when file does not exist", () => {
    test("should throw an error", async () => {
      await expect(loadConfig(join(fixturePath, "nowhere"))).rejects.toThrow(
        "ENOENT"
      );
    });
  });

  describe("when config is invalid", () => {
    test("should throw ValidationError", async () => {
      await expect(
        loadConfig(join(fixturePath, "invalid", "kosko.toml"))
      ).rejects.toThrow(ValidationError);
    });
  });
});

describe("searchConfig", () => {
  test("should return the config when kosko.toml exists", async () => {
    const config = await searchConfig(join(fixturePath, "toml"));
    expect(config).toMatchSnapshot();
  });

  test("should return an empty object when kosko.toml does not exist", async () => {
    const config = await searchConfig(join(fixturePath, "nowhere"));
    expect(config).toEqual({});
  });

  test("should throw ValidationError when config is invalid", async () => {
    const cwd = join(fixturePath, "invalid");
    await expect(searchConfig(cwd)).rejects.toThrow(ValidationError);
  });
});

describe("getConfig", () => {
  test("should return an empty config when config is empty", () => {
    expect(getConfig({}, "dev")).toEqual({
      components: [],
      require: [],
      loaders: [],
      plugins: []
    });
  });

  test("should return the global config when env is not defined", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      plugins: [{ name: "a" }]
    };

    expect(getConfig(input, "dev")).toEqual(input);
  });

  test("should return the global config when env is defined but the key does not exist", () => {
    const globalConf = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      plugins: [{ name: "a" }]
    };
    const input = {
      ...globalConf,
      environments: {
        prod: {
          components: ["aaa"],
          require: ["bbb"],
          loaders: ["ccc"],
          plugins: [{ name: "b" }]
        }
      }
    };

    expect(getConfig(input, "dev")).toEqual(globalConf);
  });

  test("should return the merge of global and environment config when env is defined and the key exists", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      plugins: [{ name: "a" }],
      environments: {
        dev: {
          components: ["aaa"],
          require: ["bbb"],
          loaders: ["ccc"],
          plugins: [{ name: "b" }]
        }
      }
    };

    expect(getConfig(input, "dev")).toEqual({
      components: ["foo", "aaa"],
      require: ["bar", "bbb"],
      loaders: ["baz", "ccc"],
      plugins: [{ name: "a" }, { name: "b" }]
    });
  });

  test("should return the merge of multiple envs when env is an array", () => {
    const input = {
      components: ["foo"],
      require: ["bar"],
      loaders: ["baz"],
      plugins: [{ name: "a" }],
      environments: {
        a: {
          components: ["aa"],
          require: ["ab"],
          loaders: ["ac"],
          plugins: [{ name: "ad" }]
        },
        c: {
          components: ["ca"],
          require: ["cb"],
          loaders: ["cc"],
          plugins: [{ name: "cd" }]
        }
      }
    };

    expect(getConfig(input, ["a", "b", "c"])).toEqual({
      components: ["foo", "aa", "ca"],
      require: ["bar", "ab", "cb"],
      loaders: ["baz", "ac", "cc"],
      plugins: [{ name: "a" }, { name: "ad" }, { name: "cd" }]
    });
  });
});
