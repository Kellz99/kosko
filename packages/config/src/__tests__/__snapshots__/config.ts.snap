// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig when file exists should load the config 1`] = `
Object {
  "baseEnvironment": "m",
  "components": Array [
    "b",
    "c",
  ],
  "environments": Object {
    "dev": Object {
      "components": Array [
        "f",
        "g",
      ],
      "require": Array [
        "d",
        "e",
      ],
    },
    "prod": Object {
      "components": Array [
        "i",
        "j",
      ],
      "require": Array [
        "h",
      ],
    },
  },
  "extensions": Array [
    "js",
    "json",
  ],
  "paths": Object {
    "environment": Object {
      "component": "l",
      "global": "k",
    },
  },
  "require": Array [
    "a",
  ],
}
`;

exports[`searchConfig succeed when config is at kosko.toml should load the config 1`] = `
Object {
  "baseEnvironment": "m",
  "components": Array [
    "b",
    "c",
  ],
  "environments": Object {
    "dev": Object {
      "components": Array [
        "f",
        "g",
      ],
      "require": Array [
        "d",
        "e",
      ],
    },
    "prod": Object {
      "components": Array [
        "i",
        "j",
      ],
      "require": Array [
        "h",
      ],
    },
  },
  "extensions": Array [
    "js",
    "json",
  ],
  "paths": Object {
    "environment": Object {
      "component": "l",
      "global": "k",
    },
  },
  "require": Array [
    "a",
  ],
}
`;
